File: Solved/problem764.py
import math
from sys import stdout
from time import time
from typing import Callable

MODULO = 10 ** 9
MAX = 10 ** 16
MAX_SAFE_INTEGER = 2**53 - 1

class Tracer:
    def __init__(self, prefix: str, enabled: bool) -> None:
        self.enabled = enabled
        self.lastPrinted = time()-1
        self.lastValue = ''
        self.prefix = prefix

    def print(self, callback: Callable[[], str]) -> None:
        if not self.enabled:
            return
        value = f'{callback()}'
        if value == self.lastValue:
            return
        now = time()
        if now-self.lastPrinted < 1:
            return
        stdout.write(f'\r{self.prefix}: {value.rjust(len(self.lastValue))} ')
        self.lastPrinted = now
        self.lastValue = value

    def clear(self):
        spaces = ''
        stdout.write(f'\r{spaces.rjust(len(self.prefix) + 4 + len(self.lastValue))}\r')

def sqrt(n):
    r = int(math.sqrt(n))+1
    while r*r <= n:
        r += 1
    return r-1
    
def getEvenSquares(max):
    i = 2
    while True:
        s = i*i
        if i > max:
            return

        yield s
        i += 2

def getSpecialValues(max):
    i = 1
    while True:
        value = 8 * i * (i - 1) + 2
        if value > max:
            return
        yield value
        i += 1

def add(total, x, y, z):
    if math.gcd(math.gcd(x, y), z) == 1:
            return (total + x + y + z) % MODULO
    return total
        
def S(max, trace = False):
    total = 0

    maxN = sqrt(max // 2)
    squares = [ s for s in getEvenSquares(sqrt(max)) ]
    specialValues = [ s for s in getSpecialValues(sqrt(max)) ]

    # special with squares

    tracer = Tracer('m is square', trace)
    offset = 0
    for i, n in enumerate(specialValues):
        if n > maxN:
            break

        tracer.print(lambda: maxN-n)

        n2 = n*n
        maxM = sqrt(max - n2)

        for j in range(i+offset, len(squares)):
            m = squares[j]
            if m > maxM:
                break

            if m < n:
                offset += 1 
                continue

            m2 = m*m

            a = m2 - n2
            b = 2 * m * n
            c = n2 + m2
            total = add(total, a // 4, sqrt(b), c)

    tracer.clear()

    # n are squares

    tracer = Tracer('n is square', trace)
    offset = 0
    for i, n in enumerate(squares):
        if n > maxN: 
            break

        tracer.print(lambda: maxN-n)

        n2 = n*n
        maxM = sqrt(max - n2)

        for j in range(offset, len(specialValues)):
            m = specialValues[j]
            if m > maxM:
                break

            if m < n:
                offset += 1
                continue

            m2 = m*m
    
            a = m2 - n2
            b = 2 * m * n
            c = n2 + m2

            if a % 4 == 0:
                total = add(total, a // 4, sqrt(b), c)

    tracer.clear()

    # n multiple of 4 and n+m is square

    tracer = Tracer('n+m is square', trace)

    for n in range(4, maxN+1, 4):

        tracer.print(lambda: maxN-n)

        n2 = n*n
        maxM = sqrt(max - n2)

        for q in range(sqrt(n)+1, maxM+1):
            q2 = q*q
            m = q2 - n
            if m > maxM:
                break

            if m < n: 
                continue

            m2 = m*m

            a = m2 - n2
            b = 2 * m * n
            c = n2 + m2

            aa = sqrt(a)
            if aa*aa == a:
                total = add(total, b // 4, aa, c)

    tracer.clear()

    return total

start = time()
assert S(100) == 81
assert S(10000) == 112851
assert S(1E7) == 248876211
assert S(1E8) == 42501751
assert S(1E9) == 258133745
assert S(1E12) == 397047378
assert S(1E13, True) == 769855408
end = time()

print(F"Tests executed in { math.floor( (end-start) * 1000 ) / 1000 } seconds")

start = time()
answer = S(MAX, True)
end = time()

print(F"Asnwer is {answer} executed in { math.floor( (end-start )* 1000 ) / 1000 } seconds")

File: Solved/problem764.js
const assert = require('assert');
const timeLogger = require('tools/timeLogger');
const Tracer = require('tools/tracer');

const MODULO = 10n ** 9n;
const MAX = 10n ** 16n;

require('tools/bigintHelper');
require('tools/numberHelper');

// a=3 , b=4, c=20

// a = m^2 - n^2 = (m+n)(m-n)
// b = 2*n*m    
// c = n^2 + m^2

function getEvenSquares(max)
{
    const values = [];

    for(let i = 2n; ; i += 2n) {
        const s = i*i;

        if (i > max) {
            break;
        }

        values.push(s);
    }

    return values;
}

function getSpecialValues(max)
{
    const special = [];
    for(let i = 1n; ; i++) {
        const value = 8n * i * (i - 1n) + 2n;
        if (value > max) {
            break;
        }
        special.push(value); 
    }
    return special;
}

function generateTriangles(max, trace, callback)
{
    max = BigInt(max);

    const maxN = (max / 2n).sqrt();
    const squares = getEvenSquares(max.sqrt());
    const specialValues = getSpecialValues(max.sqrt());

    // special with squares

    let tracer = new Tracer(1, trace, 'Square m:');

    let a, b, c;
    let offset = 0;

    for(let i = 0; i < specialValues.length; i++) {
        const n = specialValues[i];

        if (n > maxN) {
            break;
        }

        tracer.print(_ => maxN - n);

        const n2 = n*n;

        const maxM = (max - n2).sqrt();

        for(let j = i+offset; ; j++) {
            const m = squares[j];
            if (m > maxM) {
                break;
            }

            if (m < n) {
                offset++;
                continue;
            }

            const m2 = m*m;

            a = m2 - n2;
            b = 2n*m*n;
            c = n2 + m2;
            callback(a / 4n, b.sqrt(), c);
        }
    }
    tracer.clear();

    // n are squares

    tracer = new Tracer(1, trace, 'Square n:');

    offset = 0;
    for(let i = 0; ; i++) {
        const n = squares[i];
        
        if (n > maxN) {
            break;
        }

        tracer.print(_ => maxN - n);

        const n2 = n*n;
        const maxM = (max - n2).sqrt();

        for(let j = offset; j < specialValues.length; j++) {
            const m = specialValues[j];
            if (m > maxM) {
                break;
            }
            if (m < n) {
                offset++;
                continue;
            }

            const m2 = m*m;
    
            a = (m2 - n2);
            b = 2n * m * n;
            c = n2 + m2;

            if (a % 4n === 0n) {
                callback(a / 4n, b.sqrt(), c);
            }
        }
    }
    tracer.clear();

    // n multiple of 4 and n+m is square

    tracer = new Tracer(1, trace, 'Square n+m:');

    for (let n = 4; n <= maxN ; n += 4) {
        tracer.print(_ => maxN - BigInt(n));

        const n2 = n*n;
        const maxM = (max - BigInt(n2)).sqrt();

        for (let q = n.sqrt()+1; ; q++) {
            const q2 = q*q;

            const m = q2 - n;
            if (m > maxM) {
                break;
            }

            if (m < n) {
                continue;
            }

            let m2 = m*m;

            a = m2 - n2;
            b = 2 * m * n;
            c = n2 + m2;

            if (m2 >= Number.MAX_SAFE_INTEGER || b > Number.MAX_SAFE_INTEGER || c > Number.MAX_SAFE_INTEGER) {
                m2 = BigInt(m)*BigInt(m);
                a = m2 - BigInt(n2);
                b = 2n * BigInt(m) * BigInt(n);
                c = BigInt(n2) + m2;

                const aa = a.sqrt();

                if (aa*aa === a) {
                    callback(b / 4n, aa, c);
                }
            } else {
                const aa = a.sqrt();

                if (aa*aa === a) {
                    callback(b / 4, aa, c);
                }
            }
        }
    }
    tracer.clear();
}

function S(N, trace) 
{
    let total = 0;

    const MOD = Number(MODULO);

    generateTriangles(N, trace, (x, y, z) => {
        if (z <= Number.MAX_SAFE_INTEGER) {
            x = Number(x);
            y = Number(y);
            z = Number(z);

            if (x.gcd(y).gcd(z) === 1) {
                const xyz = (x % MOD) + (y % MOD) + (z % MOD);
                total = (total + xyz) % MOD;
            }
        } else if (x.gcd(y).gcd(z) === 1n) {
            const xyz = Number((x+y+z) % MODULO);
            total = (total + xyz) % MOD;
        }
    });

    return total;
}

assert.strictEqual(S(100), 81);
assert.strictEqual(S(10000), 112851);
assert.strictEqual(timeLogger.wrap('S(1E7)' , _ => S(1E7)), 248876211);
assert.strictEqual(timeLogger.wrap('S(1E8)' , _ => S(1E8)), 42501751);
assert.strictEqual(timeLogger.wrap('S(1E9)' , _ => S(1E9)), 258133745);
assert.strictEqual(timeLogger.wrap('S(1E12)' , _ => S(1E12)), 397047378);

console.log('Tests passed');

const answer = timeLogger.wrap('', _ => S(MAX, true));
console.log(`Answer is ${answer}`);